File: day8/index1.js
const { readLines, parseInt } = require("../lib");

const parse = line => {
  const m = line.match(
    /^(\w+) (inc|dec) (-?\d+) if (\w+) (<|>|<=|>=|==|!=) (-?\d+)$/
  );
  if (!m) {
    throw new Error(`WTF ${line}`);
  }
  return {
    reg: m[1],
    dec: m[2] === "dec",
    val: parseInt(m[3]),
    condReg: m[4],
    cond: m[5],
    condVal: parseInt(m[6])
  };
};

const data = readLines("./input").map(parse);

const reg = {};

const get = name => reg[name] || 0;

const update = (name, dec, val) => {
  reg[name] = get(name) + val * (dec ? -1 : 1);
};

const condHolds = (name, cond, val) => {
  const x = get(name);
  switch (cond) {
    case "<":
      return x < val;
    case "<=":
      return x <= val;
    case ">":
      return x > val;
    case ">=":
      return x >= val;
    case "==":
      return x == val;
    case "!=":
      return x != val;
    default:
      throw new Error(`WTF unknown cond ${cond} for ${name}`);
  }
};

data.forEach(({ reg, dec, val, condReg, cond, condVal }) => {
  if (condHolds(condReg, cond, condVal)) {
    update(reg, dec, val);
  }
});

console.log(Math.max(...Object.values(reg)));

File: clojure/day08/index1.clj
(require '[clojure.string :as s])

(def parse-int #(Integer/parseInt %))

(defn split [p s] (s/split s p))

(defn parse [line]
  (def re #"^(\w+) (inc|dec) (-?\d+) if (\w+) (<|>|<=|>=|==|!=) (-?\d+)$")
  (def m (re-matches re line))
  (if-not m 
    (throw (ex-info "WTF" {:line line}))
    (let [
      [_ name op val cnd-name cnd cnd-val] m
    ] {
      :name name
      :dec (= op "dec")
      :val (parse-int val)
      :cnd-name cnd-name
      :cnd cnd
      :cnd-val (parse-int cnd-val)
    })))

(def ops {
  "<"   <
  "<="  <=
  ">"   >
  ">="  >=
  "=="  ==
  "!="  not=
})

(def data (->> "./input" slurp s/trim-newline s/split-lines (map parse)))

(defn get-val [reg name] (get reg name 0))

(defn update-val [reg name dec val] (let [
  old-val (get-val reg name)
  mul (if dec -1 1)
  new-val (+ old-val (* val mul))
] (assoc reg name new-val)))

(defn cond-holds [reg name cnd val]
  (let [
    reg-val (get-val reg name)
    op (get ops cnd)
  ] (if-not op
    (throw (ex-info "WTF unknown cond" {:cnd cnd :name name}))
    (op reg-val val))))

(def registers (loop [
  registers {}
  rem-data data
] (if (== 0 (count rem-data)) registers
  (let [
    d (first rem-data)
    rem-data' (rest rem-data)
    {:keys [name dec val cnd-name cnd cnd-val]} d
    holds (cond-holds registers cnd-name cnd cnd-val)
    registers' (if-not holds registers (update-val registers name dec val))
  ] (
    recur registers' rem-data'
  )))))

(def res (apply max (vals registers)))

(println res)